File: day01/day01.rs
use std::fs;

fn parse_line(line: &str) -> i32 {
    let direction = line.chars().next().unwrap();
    let value: i32 = line[1..].parse().unwrap();
    if direction == 'L' {
        -value
    } else {
        value
    }
}

fn count_crossings(dial: i32, movement: i32) -> i32 {
    let full_rotations = movement.abs() / 100;
    let sign = if movement >= 0 { 1 } else { -1 };
    let remainder = movement - sign * 100 * full_rotations;
    let new_pos = dial + remainder;

    let boundary_crossing = if remainder < 0 && dial > 0 && new_pos <= 0 {
        1
    } else if remainder > 0 && new_pos > 99 {
        1
    } else {
        0
    };

    full_rotations + boundary_crossing
}

fn normalize_dial(dial: i32) -> i32 {
    ((dial % 100) + 100) % 100
}

fn main() {
    let content = fs::read_to_string("input.txt").expect("Failed to read input.txt");

    let (_, zeros, crossings) = content.lines().filter(|line| !line.is_empty()).fold(
        (50, 0, 0),
        |(dial, zeros, crossings), line| {
            let movement = parse_line(line);
            let new_dial = normalize_dial(dial + movement);
            let new_zeros = if new_dial == 0 { zeros + 1 } else { zeros };
            let new_crossings = crossings + count_crossings(dial, movement);
            (new_dial, new_zeros, new_crossings)
        },
    );

    println!("{}", zeros);
    println!("{}", crossings);
}

File: day01/Day01.cs
var lines = File.ReadAllLines("input.txt").Where(l => !string.IsNullOrEmpty(l)).ToArray();

int ParseLine(string line)
{
    var direction = line[0];
    var value = int.Parse(line[1..]);
    return direction == 'L' ? -value : value;
}

int CountCrossings(int dial, int movement)
{
    var fullRotations = Math.Abs(movement) / 100;
    var sign = movement >= 0 ? 1 : -1;
    var remainder = movement - sign * 100 * fullRotations;
    var newPos = dial + remainder;

    var boundaryCrossing =
        (remainder < 0 && dial > 0 && newPos <= 0) ? 1 :
        (remainder > 0 && newPos > 99) ? 1 : 0;

    return fullRotations + boundaryCrossing;
}

int NormalizeDial(int dial) => ((dial % 100) + 100) % 100;

var dial = 50;
var zeros = 0;
var crossings = 0;

foreach (var line in lines)
{
    var movement = ParseLine(line);
    var newDial = NormalizeDial(dial + movement);
    if (newDial == 0) zeros++;
    crossings += CountCrossings(dial, movement);
    dial = newDial;
}

Console.WriteLine(zeros);
Console.WriteLine(crossings);
